src/platform/w3dengine/client/w3dstatuscircle.cpp

```diff
@@ -0,0 +1,290 @@
+ *            LICENSE
+ */
+#include "w3dstatuscircle.h"
+#include "dx8indexbuffer.h"
+#include "dx8vertexbuffer.h"
+#include "dx8wrapper.h"
+#include "gamelogic.h"
+#include "globaldata.h"
+#include "scriptengine.h"
+#include "vertmaterial.h"
+
+int W3DStatusCircle::m_diffuse = 0xFF;
+bool W3DStatusCircle::m_needUpdate;
+
+W3DStatusCircle::W3DStatusCircle() :
+    m_indexBuffer(nullptr), m_vertexMaterialClass(nullptr), m_vertexBufferCircle(nullptr), m_vertexBufferScreen(nullptr)
+{
+}
+
+W3DStatusCircle::W3DStatusCircle(const W3DStatusCircle &src)
+{
+    *this = src;
+}
+
+W3DStatusCircle &W3DStatusCircle::operator=(const W3DStatusCircle &that)
+{
+    return *this;
+}
+
+W3DStatusCircle::~W3DStatusCircle()
+{
+    Free_Map_Resources();
+}
+
+bool W3DStatusCircle::Cast_Ray(RayCollisionTestClass &raytest)
+{
+    return false;
+}
+
+void W3DStatusCircle::Get_Obj_Space_Bounding_Sphere(SphereClass &sphere) const
+{
+    Vector3 v(500.0f, 500.0f, 0.0f);
+    sphere.Init(v, v.Length());
+}
+
+void W3DStatusCircle::Get_Obj_Space_Bounding_Box(AABoxClass &box) const
+{
+    box.Init(Vector3(0.0f, 0.0f, 0.0f), Vector3(1000.0f, 1000.0f, 1000.0f));
+}
+
+int W3DStatusCircle::Class_ID() const
+{
+    return CLASSID_UNKNOWN;
+}
+
+RenderObjClass *W3DStatusCircle::Clone() const
+{
+    return new W3DStatusCircle(*this);
+}
+
+int W3DStatusCircle::Free_Map_Resources()
+{
+    Ref_Ptr_Release(m_indexBuffer);
+    Ref_Ptr_Release(m_vertexBufferScreen);
+    Ref_Ptr_Release(m_vertexBufferCircle);
+    Ref_Ptr_Release(m_vertexMaterialClass);
+    return 0;
+}
+
+int W3DStatusCircle::Init_Data()
+{
+    m_needUpdate = true;
+    Free_Map_Resources();
+    m_numTriangles = 20;
+    m_indexBuffer = new DX8IndexBufferClass(3 * m_numTriangles, DX8IndexBufferClass::USAGE_DEFAULT);
+    IndexBufferClass::WriteLockClass lock(m_indexBuffer, 0);
+    unsigned short *indices = lock.Get_Index_Array();
+
+    for (int i = 0; i < 3 * m_numTriangles; i += 3) {
+        indices[0] = i;
+        indices[1] = i + 1;
+        indices[2] = i + 2;
+        indices += 3;
+    }
+
+    m_vertexBufferCircle =
+        new DX8VertexBufferClass(DX8_FVF_XYZDUV1, 3 * m_numTriangles, DX8VertexBufferClass::USAGE_DEFAULT, 0);
+    m_vertexBufferScreen = new DX8VertexBufferClass(DX8_FVF_XYZDUV1, 6, DX8VertexBufferClass::USAGE_DEFAULT, 0);
+    m_vertexMaterialClass = VertexMaterialClass::Get_Preset(VertexMaterialClass::PRELIT_DIFFUSE);
+    m_shaderClass = ShaderClass(0x984B7);
+    return 0;
+}
+
+int W3DStatusCircle::Update_Circle_VB()
+{
+    if (!m_vertexBufferCircle) {
+        return -1;
+    }
+
+    m_needUpdate = false;
+    VertexBufferClass::WriteLockClass lock(m_vertexBufferCircle, 0);
+    VertexFormatXYZDUV1 *vertices = (VertexFormatXYZDUV1 *)lock.Get_Vertex_Array();
+    float z = 0.0f;
+    float x = 0.02f;
+    int diffuse = m_diffuse + 0x7F000000;
+    int count = m_numTriangles;
+    float angle = 0.0f;
+    float deltaAngle = 6.2831855f / (float)count;

```

Comment:
pi * 2

Reply:
should be using the define
and delta_angle